### PR TITLE
Use previous lookup of database on startup and remove point lookups.

### PR DESCRIPTION
**Pull request by jschlather**
Addresses some of #1226.

If there's any extra testing that I need to do for this, I'll probably need some direction on where to write it.

### CODE REVIEW

**Comment by fommil** [comment]
wow, that was fast! OK, I need to sit down and really study this because the original code was terrible and prone to failures.

**Comment by fommil** [comment]
I'm reviewing this now, sorry for the delay. Looks like one of our CI workers ran out of memory, so restarting.

**Comment by fommil** [comment]
heh, just trying out the git `hub` module for the first time. I typed this to get the PR applied locally

```
hub am https://github.com/ensime/ensime-server/pull/1228
```

**Comment by fommil** [comment]
this is looking really great, but I have a few tiny comments to make. Instead of bothering you with them, I'll just squash my changes onto your commit if that's ok.

File: core/src/main/scala/org/ensime/indexer/SearchService.scala [hunk 1]
```diff
@@ -89,37 +89,39 @@ class SearchService(
...
       }
+    }
 
     // index all the given bases and return number of rows written
-    def indexBases(bases: Set[FileObject]): Future[Int] = {
-      log.info("indexBases")
-      Future.sequence(bases.map(indexBase)).map(_.flatten.sum)
+    def indexBases(bases: Set[FileObject], checks: Seq[FileCheck]): Future[Int] = {
+      log.info("Indexing bases...")
+      val checksLookup: Map[String, FileCheck] = checks.map(check => (check.filename -> check)).toMap
```
Comment (fommil) [inline core/src/main/scala/org/ensime/indexer/SearchService.scala]: prefer `breakOut` to `toMap` and friends https://skillsmatter.com/skillscasts/6939-optimising-scala-for-fun-and-profit

File: core/src/main/scala/org/ensime/indexer/SearchService.scala [hunk 2]
```diff
@@ -89,37 +89,39 @@ class SearchService(
...
+      val basesWithChecks: Set[(FileObject, Option[FileCheck])] = bases.map { base =>
```
Comment (fommil) [inline core/src/main/scala/org/ensime/indexer/SearchService.scala]: is also a Map

**Comment by jschlather** [comment]
Sure, whatever works.

**Comment by fommil** [comment]
FYI, I'll squash this onto your commit and raise a fresh PR. git history will say it was all you.

```diff
@@ -10,6 +10,7 @@ import org.ensime.indexer.DatabaseService._
 import org.ensime.util.file._
 
 //import scala.concurrent.ExecutionContext.Implicits.global
+import scala.collection.breakOut
 import scala.concurrent._
 import scala.concurrent.duration._
 
@@ -117,11 +118,16 @@ class SearchService(
     // index all the given bases and return number of rows written
     def indexBases(bases: Set[FileObject], checks: Seq[FileCheck]): Future[Int] = {
       log.info("Indexing bases...")
-      val checksLookup: Map[String, FileCheck] = checks.map(check => (check.filename -> check)).toMap
-      val basesWithChecks: Set[(FileObject, Option[FileCheck])] = bases.map { base =>
+      val checksLookup: Map[String, FileCheck] = checks.map { check =>
+        (check.filename -> check)
+      }(breakOut)
+      val basesWithChecks: Map[FileObject, Option[FileCheck]] = bases.map { base =>
         (base, checksLookup.get(base.getName().getURI()))
+      }(breakOut)
+      val work = basesWithChecks.map {
+        case (file, check) => indexBase(file, check)
       }
-      Future.sequence(basesWithChecks.map { case (file, check) => indexBase(file, check) }).map(_.flatten.sum)
+      Future.sequence(work).map(_.flatten.sum)
     }
 
     def commitIndex(): Future[Unit] = Future {
```

**Comment by fommil** [comment]
unless you just apply that on here and force push.

**Comment by fommil** [comment]
hold on, one more small change.. we can create `FileObject`s in the first bit, I've got a more efficient way to construct the Map.

**Comment by fommil** [comment]
one less collection... should keep the heap requirements down (these collections can get really big, so its worth optimising...)

```diff
diff --git a/core/src/main/scala/org/ensime/indexer/SearchService.scala b/core/src/main/scala/org/ensime/indexer/SearchService.scala
index 53da84b..078932c 100644
--- a/core/src/main/scala/org/ensime/indexer/SearchService.scala
+++ b/core/src/main/scala/org/ensime/indexer/SearchService.scala
@@ -10,6 +10,7 @@ import org.ensime.indexer.DatabaseService._
 import org.ensime.util.file._
 
 //import scala.concurrent.ExecutionContext.Implicits.global
+import scala.collection.breakOut
 import scala.concurrent._
 import scala.concurrent.duration._
 
@@ -117,11 +118,15 @@ class SearchService(
     // index all the given bases and return number of rows written
     def indexBases(bases: Set[FileObject], checks: Seq[FileCheck]): Future[Int] = {
       log.info("Indexing bases...")
-      val checksLookup: Map[String, FileCheck] = checks.map(check => (check.filename -> check)).toMap
-      val basesWithChecks: Set[(FileObject, Option[FileCheck])] = bases.map { base =>
-        (base, checksLookup.get(base.getName().getURI()))
+      val known: Map[FileObject, FileCheck] = checks.flatMap { check =>
+        val f = check.file
+        if (bases(f)) Some(check.file -> check)
+        else None
+      }(breakOut)
+      val work = bases.map {
+        file => indexBase(file, known.get(file))
       }
-      Future.sequence(basesWithChecks.map { case (file, check) => indexBase(file, check) }).map(_.flatten.sum)
+      Future.sequence(work).map(_.flatten.sum)
     }
 
     def commitIndex(): Future[Unit] = Future {
```

**Comment by fommil** [comment]
I'll send it as a small PR. Really great stuff, thank you!

**Comment by jschlather** [comment]
Cool, just catching up. Hadn't heard of breakout before, thanks for the information. Glad to help out.

**Comment by rorygraves** [comment]
Good stuff

**Pull request merged**